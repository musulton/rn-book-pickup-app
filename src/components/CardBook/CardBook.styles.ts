import {ImageStyle, StyleSheet, TextStyle, ViewStyle} from "react-native";

import {COLORS, SIZES} from "../../constants";

interface Styles {
    container: ViewStyle
    leftSection: ViewStyle
    rightSection: ViewStyle
    image: ImageStyle
    title: TextStyle
    author: TextStyle
    edition: TextStyle
    pickupDate: TextStyle
}

const styles = StyleSheet.create<Styles>({
    container: {
        display: "flex",
        flexDirection: "row",
        alignItems: "center",
        paddingHorizontal: SIZES.padding2,
        marginVertical: SIZES.padding2
    },
    leftSection: {
        marginRight: 5,
        resizeMode: "stretch",
    },
    rightSection: {
        flex: 1,
        display: "flex",
        marginLeft: SIZES.padding2,
        alignContent: "center",
        justifyContent: "center"
    },
    image: {
        height: 100,
        width: 70,
        borderRadius: SIZES.borderRadius,
        resizeMode: "cover"
    },
    title: {
        fontSize: 18,
        fontWeight: "500"
    },
    author: {
        color: COLORS.lightGray,
        marginVertical: 5
    },
    edition: {
        fontSize: 16,
        fontWeight: "500"
    },
    pickupDate: {
        fontSize: 16,
        marginTop: 5,
    }
})

export default styles
