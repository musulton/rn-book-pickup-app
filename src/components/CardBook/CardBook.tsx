import React from "react";
import {TouchableOpacity, Text, Image, View} from "react-native";

import styles from "./CardBook.styles";
import {Book} from "../../types";

interface CardBookProps extends Book {
    onPress?: () => void
}

interface LeftSectionProps {
    coverImg?: string
}

const LeftSection: React.FC<LeftSectionProps> = (props) => {
    return (
        <View style={styles.leftSection}>
            <Image
                source={{uri: props.coverImg}}
                style={styles.image}
            />
        </View>
    )
}

const RightSection: React.FC<Book> = (props) => {
    return (
        <View style={styles.rightSection}>
            <Text style={styles.title} numberOfLines={2}>{props.title}</Text>
            <Text style={styles.author} numberOfLines={1}>{props.author}</Text>
            <Text style={styles.edition} numberOfLines={1}>{props.edition}</Text>
            {props.pickupDate && (
                <Text style={styles.pickupDate}>
                    <>Book Date: {props.pickupDate}</>
                </Text>
            )}
        </View>
    )
}

const CardBook: React.FC<CardBookProps> = (props: CardBookProps) => {
    return (
        <TouchableOpacity
            style={styles.container}
            onPress={props.onPress}
            disabled={!props.onPress}
        >
            <LeftSection coverImg={props.coverImg} />
            <RightSection title={props.title} author={props.author} edition={props.edition} pickupDate={props.pickupDate} />
        </TouchableOpacity>
    )
}

export default React.memo(CardBook)
