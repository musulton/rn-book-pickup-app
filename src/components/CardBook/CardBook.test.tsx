import {fireEvent, render, screen} from "@testing-library/react-native";

import {book} from "../../fixtures/testData";
import CardBook from "./CardBook";

describe("<CardBook />", () => {
    const mockProps = {
        onPress: jest.fn(),
        ...book,
        coverImg: "google.com"
    }

    it("should render <CardBook /> correctly", () => {
        const props = {...mockProps}
        render(<CardBook {...props} />)

        expect(screen.toJSON()).toMatchSnapshot();
    })

    it("should render <CardBook /> correctly with pickupDate", () => {
        const props = {
            ...mockProps,
            pickupDate: "12 Mei 2023"
        }

        render(<CardBook {...props} />)

        expect(screen.toJSON()).toMatchSnapshot();
    })

    it("should call props onPress when <CardBook/> is clicked", () => {
        const props = {...mockProps}
        const {getByText} = render(<CardBook {...props} />)

        fireEvent.press(getByText("title"))
        expect(props.onPress).toHaveBeenCalled()
    })
});
