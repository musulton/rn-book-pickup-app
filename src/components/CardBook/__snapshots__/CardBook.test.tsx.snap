// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CardBook /> should render <CardBook /> correctly 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "display": "flex",
      "flexDirection": "row",
      "marginVertical": 10,
      "opacity": 1,
      "paddingHorizontal": 10,
    }
  }
>
  <View
    style={
      {
        "marginRight": 5,
        "resizeMode": "stretch",
      }
    }
  >
    <Image
      source={
        {
          "uri": "google.com",
        }
      }
      style={
        {
          "borderRadius": 6,
          "height": 100,
          "resizeMode": "cover",
          "width": 70,
        }
      }
    />
  </View>
  <View
    style={
      {
        "alignContent": "center",
        "display": "flex",
        "flex": 1,
        "justifyContent": "center",
        "marginLeft": 10,
      }
    }
  >
    <Text
      numberOfLines={2}
      style={
        {
          "fontSize": 18,
          "fontWeight": "500",
        }
      }
    >
      title
    </Text>
    <Text
      numberOfLines={1}
      style={
        {
          "color": "#64676D",
          "marginVertical": 5,
        }
      }
    >
      author
    </Text>
    <Text
      numberOfLines={1}
      style={
        {
          "fontSize": 16,
          "fontWeight": "500",
        }
      }
    >
      edition
    </Text>
  </View>
</View>
`;

exports[`<CardBook /> should render <CardBook /> correctly with pickupDate 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "display": "flex",
      "flexDirection": "row",
      "marginVertical": 10,
      "opacity": 1,
      "paddingHorizontal": 10,
    }
  }
>
  <View
    style={
      {
        "marginRight": 5,
        "resizeMode": "stretch",
      }
    }
  >
    <Image
      source={
        {
          "uri": "google.com",
        }
      }
      style={
        {
          "borderRadius": 6,
          "height": 100,
          "resizeMode": "cover",
          "width": 70,
        }
      }
    />
  </View>
  <View
    style={
      {
        "alignContent": "center",
        "display": "flex",
        "flex": 1,
        "justifyContent": "center",
        "marginLeft": 10,
      }
    }
  >
    <Text
      numberOfLines={2}
      style={
        {
          "fontSize": 18,
          "fontWeight": "500",
        }
      }
    >
      title
    </Text>
    <Text
      numberOfLines={1}
      style={
        {
          "color": "#64676D",
          "marginVertical": 5,
        }
      }
    >
      author
    </Text>
    <Text
      numberOfLines={1}
      style={
        {
          "fontSize": 16,
          "fontWeight": "500",
        }
      }
    >
      edition
    </Text>
    <Text
      style={
        {
          "fontSize": 16,
          "marginTop": 5,
        }
      }
    >
      Book Date: 
      12 Mei 2023
    </Text>
  </View>
</View>
`;
